### ISSUE TITLE
Investigate automated formatting with Prettier

### ISSUE THREAD

**Issue by ndrewr** [wevote/WebApp#1400]
Perhaps we can save time and energy spent on fixing linting/code-style mistakes if we incorporate a tool like [Prettier](https://prettier.io/) to automatically format files for us on commits.
[As suggested by the CreateReactApp team](https://github.com/facebook/create-react-app/blob/master/packages/react-scripts/template/README.md#formatting-code-automatically).

Tasks:

- install and incorporate existing WeVote eslint settings
- first run on individual files to test the formatted output for any unintentional text formatting issues
  - gather test case files? (e.g. text in components that use Popover)

**Comment by SailingSteve** [comment]
Just saw this issue.

Also see  wevote/WebApp/issues/1405

The big down side of automatic formatting of an entire codebase, is that it creates a change in every file that makes doing a diff over that "prettification" extremely difficult.  There are so many changes that diffs are a mess.  So using something like "prettier" at the beginning of a project costs nothing, and in the middle can be very expensive.  

Also the "prettier" better be flawless, since there is nothing more frustrating than a "prettier" that unintentionally breaks things (and that overrides what you put in your pull request).

See https://github.com/prettier/prettier/issues

Maybe uncleoptimus could install prettier in his IDE, and try it on one or two files as he works on them to see what happens?  Then wait a while before trying more.

**Comment by edward-ly** [comment]
I just incorporated Prettier into one of my personal projects, and here's what I know so far.

Prettier is a pretty neat tool to enforce some formatting rules as well as "beautify" files in JSON, Markdown, etc., while making sure it doesn't break any code. However, it can't fix everything because of that, and ESLint can do something similar with the `--fix` option anyway. This won't correct all of the warnings/errors it has found either, but I recommend doing this first before adding Prettier.

For instance, the file `src/js/routes/VoterGuide/OrganizationVoterGuideEdit.jsx` currently has these ESLint warnings:
```
/home/esly14/Documents/GitHub/we-vote/WebApp/src/js/routes/VoterGuide/OrganizationVoterGuideEdit.jsx
   3:10  warning  'Link' is defined but never used    no-unused-vars
   4:10  warning  'Button' is defined but never used  no-unused-vars
  86:62  warning  Missing semicolon                   semi
```
Running `./node_modules/.bin/eslint src/js/routes/VoterGuide/OrganizationVoterGuideEdit.jsx --fix` will add the missing semicolon and write it to the file, but will not do anything to the 'Link' and 'Button' variables, since we have to remove the appropriate lines manually (it turns out that 'Link' and 'Button' are imported components that are never used in `OrganizationVoterGuideEdit`).

In contrast, running `./node_modules/.bin/prettier src/js/routes/VoterGuide/OrganizationVoterGuideEdit.jsx --write` will do the same thing, but may make other changes as well depending on the options or configuration given, so I would have to run `eslint` afterwards anyway to ensure that any extra lint warnings created by Prettier are fixed.

tl;dr Manual review is still necessary, but tools like `eslint --fix` (and optionally `prettier --write`) can automate part of the process.

**Comment by uncleoptimus** [comment]
Sorry for the delay on this issue. Appreciate the comments and discussion! Working through better understanding of the tool.

Just had a good session reviewing the docs, how its plays with ESlint and popular plugins for workflow w/ ESlint. Installed the two recommended plugins as an experiment. Those are:

[eslint-plugin-prettier](https://github.com/prettier/eslint-plugin-prettier) - for when you want ESlint itself to run prettier automatically as part of its process.

[eslint-config-prettier](https://github.com/prettier/eslint-config-prettier) - for when you want to separate operations and have ESlint ignore any rule conflicts w/ Prettier.

There is also a module expressly for using Prettier to auto-format but then *maintaining desired ESlint formatting rules that conflict with Prettier*:
https://github.com/prettier/prettier-eslint


A few comments:

- Prettier def won't be perfect in its changes. We would have to identify edge cases where its formatting breaks e.g. the way text is displayed in certain circumstances. @DaleMcGrew  mentioned one such case with Popover display text

- The way I liked to use Prettier is on git-commit hook, ie it auto-runs on commits. But that wouldn't work for us since @edward-ly  pointed out we want to run ESlint afterwards which would take place before we commit

- The plugins above *or* the prettier-eslint module would address that part of the workflow. Side-note: the config plugin also has an extra command line tool that spits out rule incongruities and with that we could technically weed out all conflicts btwn the tools, if desired

- I was curious how long it would take to "prettier-ify" the codebase in the src/js folder so I just ran it on my test branch. Then I wondered if there were any immediate and obvious breaks in the app following the --write. So far, so good. Checked out the "Related Issues" and "Opinions About..." popovers and formatting holds up

- If we choose to pursue this, I think we should utilize the above plugins. In that case, ESlint hands off format/style concerns to Prettier while continuing to watch for code-quality/correctness. The workflow would be: code/debug=>run ESlint which then runs Prettier=> address reported errors
We could set up a test branch to have folks run the app thru various normal usage to see if any weirdness shows up and straighten those out one-by-one. If those prove numerous, we could elect to roll with the "prettier-eslint" module that auto-employs the --fix operation as a "final pass"


Or, we just don't use it :) I.e., case of interesting tool, bad timing.

The original premise was to see if this tool could save some back-and-forth in the cleanup-and-submit-PR phase of the workflow. As well as help out any newcomers who could be coming from whatever code-style environment. I know personally I was a bit thrown off at first by how different the styling conventions in WeVote were from the React stuff I had been doing.

**Comment by DaleMcGrew** [comment]
Closed for now until we return to this. Thank you for the research @ndrewr!
